src/dpl/src/optimization/detailed_global.cxx

```diff
@@ -63,69 +65,167 @@ void DetailedGlobalSwap::run(DetailedMgr* mgrPtr, const std::string& command)
+      tradeoff_ = std::atof(args[++i].c_str());
     }
   }
   passes = std::max(passes, 1);
   tol = std::max(tol, 0.01);
+  tradeoff_ = std::max(0.0, std::min(1.0, tradeoff_));  // Clamp to [0.0, 1.0]
 
-  int64_t last_hpwl, curr_hpwl, init_hpwl;
   uint64_t hpwl_x, hpwl_y;
-
-  curr_hpwl = Utility::hpwl(network_, hpwl_x, hpwl_y);
-  init_hpwl = curr_hpwl;
+  int64_t init_hpwl = Utility::hpwl(network_, hpwl_x, hpwl_y);
   if (init_hpwl == 0) {
     return;
   }
+
+  // Store original displacement limits for restoration later
+  int orig_disp_x, orig_disp_y;
+  mgr_->getMaxDisplacement(orig_disp_x, orig_disp_y);
+  
+  // Get chip dimensions for unleashing the optimizer
+  const int chip_width = arch_->getMaxX().v - arch_->getMinX().v;
+  const int chip_height = arch_->getMaxY().v - arch_->getMinY().v;
+
+  mgr_->getLogger()->info(DPL, 906, "Starting two-pass congestion-aware global swap optimization (tradeoff={:.1f})", tradeoff_);
+
+  // PASS 1: HPWL Profiling Pass
+  mgr_->getLogger()->info(DPL, 907, "Pass 1: HPWL profiling to determine budget");
+  
+  // Clear journal to ensure clean state tracking for profiling pass
+  mgr_->getJournal().clear();
+  
+  is_profiling_pass_ = true;
+  congestion_weight_ = 0.0;  // Pure HPWL optimization
+  
+  int64_t last_hpwl, curr_hpwl = init_hpwl;
   for (int p = 1; p <= passes; p++) {
     last_hpwl = curr_hpwl;
-
-    // XXX: Actually, global swapping is nothing more than random
-    // greedy improvement in which the move generating is done
-    // using this object to generate a target which is the optimal
-    // region for each candidate cell.
     globalSwap();
-
     curr_hpwl = Utility::hpwl(network_, hpwl_x, hpwl_y);
-
-    mgr_->getLogger()->info(DPL,
-                            306,
-                            "Pass {:3d} of global swaps; hpwl is {:.6e}.",
-                            p,
-                            (double) curr_hpwl);
-
-    if (last_hpwl == 0
-        || std::abs(curr_hpwl - last_hpwl) / (double) last_hpwl <= tol) {
+    
+    mgr_->getLogger()->info(DPL, 316, "Profiling pass {:d}; hpwl is {:.6e}.", p, (double) curr_hpwl);
+    
+    if (last_hpwl == 0 || std::abs(curr_hpwl - last_hpwl) / (double) last_hpwl <= tol) {
       break;
     }
   }
-  double curr_imp = (((init_hpwl - curr_hpwl) / (double) init_hpwl) * 100.);
-  mgr_->getLogger()->info(DPL,
-                          307,
-                          "End of global swaps; objective is {:.6e}, "
-                          "improvement is {:.2f} percent.",
-                          (double) curr_hpwl,
-                          curr_imp);
+  
+  // Calculate budget: allow 10% degradation from optimal HPWL
+  double optimal_hpwl = curr_hpwl;
+  budget_hpwl_ = optimal_hpwl * 1.10;
+  
+  mgr_->getLogger()->info(DPL, 908, 
+                         "Profiling complete. Optimal HPWL={:.2f}, Budget HPWL={:.2f} (+10%)", 
+                         optimal_hpwl, budget_hpwl_);
```

Comment:
Aside from hard-coding 1.10 I wouldn't duplicate that choice in the message when you can easily compute the (+10%) from budget_hpwl_  & optimal_hpwl (eg resilient to change to 1.20).